drainer/syncer.go

```diff
@@ -550,6 +580,66 @@ func filterTable(pv *pb.PrewriteValue, filter *filter.Filter, schema *Schema) (i
 			continue
 		}
 
+		if binlogFilter != nil {
+			var (
+				insertIdx = 0
+				deleteIdx = 0
+				updateIdx = 0
+
+				filteredIdx       = 0
+				filteredInsertIdx = 0
+				filteredDeleteIdx = 0
+				filteredUpdateIdx = 0
+			)
+			for i, tp := range mutation.Sequence {
+				var et bf.EventType
+				switch tp {
+				case binlog.MutationType_Insert:
+					et = bf.InsertEvent
+					insertIdx++
+				case binlog.MutationType_Update:
+					et = bf.UpdateEvent
+					updateIdx++
+				case binlog.MutationType_DeleteRow:
+					et = bf.DeleteEvent
+					deleteIdx++
+				default:
+					err = errors.Errorf("unknown mutation type: %v", tp)
+					return
+				}
+
+				needSkip, err := skipByFilter(binlogFilter, schemaName, tableName, et, "")
+				if err != nil {
+					return false, errors.Trace(err)
+				} else if needSkip {
+					continue
+				}
+				mutation.Sequence[filteredIdx] = mutation.Sequence[i]
+				filteredIdx++
+				switch tp {
+				case binlog.MutationType_Insert:
+					mutation.InsertedRows[filteredInsertIdx] = mutation.InsertedRows[insertIdx-1]
+					filteredInsertIdx++
+				case binlog.MutationType_Update:
+					mutation.UpdatedRows[filteredUpdateIdx] = mutation.UpdatedRows[updateIdx-1]
+					filteredUpdateIdx++
+				case binlog.MutationType_DeleteRow:
+					mutation.DeletedRows[filteredDeleteIdx] = mutation.DeletedRows[deleteIdx-1]
+					filteredDeleteIdx++
+				}
+			}
+			mutation.Sequence = mutation.Sequence[0:filteredIdx]
+			if mutation.InsertedRows != nil {
+				mutation.InsertedRows = mutation.InsertedRows[0:filteredInsertIdx]
+			}
+			if mutation.UpdatedRows != nil {
+				mutation.UpdatedRows = mutation.UpdatedRows[0:filteredUpdateIdx]
+			}
+			if mutation.DeletedRows != nil {
+				mutation.DeletedRows = mutation.DeletedRows[0:filteredDeleteIdx]
+			}
```

Comment:
```suggestion
			mutation.InsertedRows = mutation.InsertedRows[0:filteredInsertIdx]
			mutation.UpdatedRows = mutation.UpdatedRows[0:filteredUpdateIdx]
			mutation.DeletedRows = mutation.DeletedRows[0:filteredDeleteIdx]
```
If rows is nil, then idx should be 0, so we can still do rows[0:0] safely. 😁